bottlerocket-os/bottlerocket
# Add support for bootstrap containers via API settings

**Issue number:**
bottlerocket-os/bottlerocket#1392

**Description of changes:**

```
5153d08c host-containers: change permissions in `/etc/host-containers`
```
This commit updates the permissions in `/etc/host-containers` from `755` to `750`

```
d59628c7 models: change ContainerImage to HostContainer
```
This commit updates the model name used to configure a host container from `ContainerImage` to `HostContainer`

```
8ce2cae3 Add support for bootstrap-containers via settings
```
This commit adds support to create bootstrap containers through the API. Bootstrap containers are host containers that can be used to setup the host during the execution of the `configured` target.

These containers are created with the prefix `boot` to prevent collisions with normal host containers. They can be setup to fail the boot process if the underlying container task exists with a non-zero status code.

```
f5e29ce9 host-ctr: support for bootstrap containers
```
This commit adds support for bootstrap containers in `host-ctr`. The  `container-type` flag was added to specify the type of container to be setup, which defaults to `host`. This allows to add support for other type of containers in the future without adding flags for each of them.

The root filesystem mounts were refactored out from `withSuperpowered` to `withRootFilesystemMounts`, since bootstrap containers require access to the underlying root filesystem without being given the superpowered treatment.

Failed container tasks will now return an error if the task exited with a non-zero status.

**Testing done:**
In `k8s 1.18`, `ecs` and `dev` aarch64:

* Run `systemctl status` to check that no units were failing
* Create `nginx` pods/containers, access them and `curl http://localhost` successfully
* Verifiy that both the control and admin host containers are running, with the proper configurations
* Execute `systemctl isolate multi-user` to verify that no services were killed/restarted
* Migrate from 1.0.7 to 1.0.8, back and forth to verify the new configurations were deleted/created
* Create a bootstrap container with the following configurations:
```toml
# user-data
[settings.bootstrap-containers.admin]
mode = "once"
source = "docker.io/arnaldo2792/sleeper:latest"
user-data = "ypXCt82h4bSlwrfKlA==" # base64 of ʕ·͡ᴥ·ʔ
``` 
```Dockerfile
# Image definition
FROM alpine
ADD bootstrap-script /
RUN chmod +x /bootstrap-script
ENTRYPOINT ["sh", "bootstrap-script"]
```

```sh
# bootstrap-script
set -e
CURRENT_DIR=/.bottlerocket/bootstrap-containers/current
BEAR_DIR=/.bottlerocket/bootstrap-containers/admin
VAR_DIR=/.bottlerocket/rootfs/var
MY_DIR=$VAR_DIR/lib/my_directory
cat $BEAR_DIR/user-data
touch $CURRENT_DIR/access
mkdir $MY_DIR
chmod -R o+r $MY_DIR
chown -R 1000:1000 $MY_DIR
sleep 120
```
* Verified that:
    * `ʕ·͡ᴥ·ʔ` appears in the journal logs
    * The directory `/var/lib/my_directory` was created, with the expected user/group ids, and permissions
    * The file `/local/bootstrap-containers/access` created through the `current` mount
    * `configured.target` was "blocked" for two minutes
    * `mark-successful-boot.service` executed successfully
    * After a reboot, the bootstrap container wasn't executed

* Update the bootstrap container's configuration with `apiclient set bootstrap-containers.admin.mode=once`, reboot and verified its execution failed because `/var/lib/my_dir` already exists.

```sh
Starting bootstrap container admin...
[8.350595] host-ctr[456]: mkdir: can't create directory '/.bottlerocket/rootfs/var/lib/my_directory': File exists
[ 8.415213] host-ctr[456]: time="2021-04-02T00:54:43Z" level=fatal msg="Container exited with non-zero status"
[FAILED] Failed to start bootstrap container admin.
```

* Update the bootstrap container's configuration with `apiclient set bootstrap-containers.admin.essential=true`, reboot and verified the boot failed:

```sh
[FAILED] Failed to start bootstrap container admin.
See 'systemctl status bootstrap-containers@admin.service' for details.
[DEPEND] Dependency failed for Bottlerocket final configuration complete.
[DEPEND] Dependency failed for Isolates multi-user.target.
[FAILED] Failed to start Isolates configured.target.
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.
